File: 22_专题/离线查询/并查集/Dynamic Graph Vertex Add Component Sum-连通分量和.py
# Dynamic Graph Vertex Add Component Sum
# 连接断开边/单点修改权值/查询连通分量和

# 0 u v 连接u v (保证u v不连接)
# 1 u v 断开u v  (保证u v连接)
# 2 u x 将u的值加上x
# 3 u 输出u所在连通块的值

# 离线查询
# n<=3e5

from typing import Tuple


class ATCRevocableUnionFindArray:
    """维护分量之和的可撤销并查集"""

    __slots__ = ("n", "parentSize", "sum", "history")

    def __init__(self, n: int):
        self.n = n
        self.parentSize = [-1] * n
        self.sum = [0] * n
        self.history = []

    def addSum(self, i: int, delta: int):
        """第i个元素的值加上v"""
        x = i
        while x >= 0:
            self.sum[x] += delta
            x = self.parentSize[x]

    def union(self, a: int, b: int) -> bool:
        x = self.find(a)
        y = self.find(b)
        if -self.parentSize[x] < -self.parentSize[y]:
            x, y = y, x
        self.history.append((x, self.parentSize[x]))
        self.history.append((y, self.parentSize[y]))
        if x == y:
            return False
        self.parentSize[x] += self.parentSize[y]
        self.parentSize[y] = x
        self.sum[x] += self.sum[y]
        return True

    def find(self, a: int) -> int:
        x = a
        while self.parentSize[x] >= 0:
            x = self.parentSize[x]
        return x

    def isConnected(self, a: int, b: int) -> bool:
        return self.find(a) == self.find(b)

    def revocate(self) -> bool:
        if not self.history:
            return False
        y, py = self.history.pop()
        x, px = self.history.pop()
        if self.parentSize[x] != px:
            self.sum[x] -= self.sum[y]
        self.parentSize[x] = px
        self.parentSize[y] = py
        return True

    def getComponentSum(self, i) -> int:
        return self.sum[self.find(i)]


class OfflineDynamicConnectivity:
    def __init__(self, n: int, q: int):
        """离线动态连通性查询

        Args:
            n (int): 顶点数
            q (int): 查询数
        """
        self.n = n
        self.q = q
        self.log = (q - 1).bit_length()
        self.size = 1 << self.log
        self.seg = [[] for _ in range(2 * self.size)]
        self.edges = []
        self.edgeId = {}
        self.remain = set()
        self.uf = ATCRevocableUnionFindArray(self.n)

    def addEdge(self, u: int, v: int, t: int):
        """時刻tに辺u-vを追加する。"""
        if u > v:
            u, v = v, u
        self.remain.add((u, v))
        self.edgeId[(u, v)] = t

    def removeEdge(self, u: int, v: int, t: int):
        """時刻tに辺u-vを削除する。"""
        if u > v:
            u, v = v, u
        self.remain.discard((u, v))
        self.edges.append((self.edgeId[(u, v)], t, (u, v)))

    def build(self):
        """SegmentTree上にinsert/eraseクエリを構築する。"""
        for e in self.remain:
            self.edges.append((self.edgeId[e], self.q, e))
        for l, r, e in self.edges:
            self._add(l, r, e)

    def run(self, func):
        """クエリを実行する。

        Args:
            func (function): 引数は時刻(kth query, 0-indexed)
        """
        stack = [1]
        while stack:
            k = stack.pop()
            if k >= 0:
                if self.size + self.q <= k:
                    continue
                stack.append(~k)
                for u, v in self.seg[k]:  # 行きがけ順
                    self.uf.union(u, v)
                if self.size <= k:
                    func(k - self.size)
                else:
                    stack.append(2 * k + 1)
                    stack.append(2 * k)
            else:
                for _ in self.seg[~k]:  # 帰りがけ順
                    self.uf.revocate()

    def _add(self, l: int, r: int, e: Tuple[int, int]):
        """SegmentTreeの[l, r)区間に辺u-vを追加する。"""
        l += self.size
        r += self.size
        while l < r:
            if l & 1:
                self.seg[l].append(e)
                l += 1
            if r & 1:
                r -= 1
                self.seg[r].append(e)
            l >>= 1
            r >>= 1


n, q = map(int, input().split())
res = []
query = []
dc = OfflineDynamicConnectivity(n, q)

for i, a in enumerate(map(int, input().split())):
    dc.uf.addSum(i, a)

for i in range(q):
    cur = tuple(map(int, input().split()))
    query.append(cur)
    if cur[0] == 0:
        _, u, v = cur
        dc.addEdge(u, v, i)
    elif cur[0] == 1:
        _, u, v = cur
        dc.removeEdge(u, v, i)


def func(k: int):
    cur = query[k]
    if cur[0] == 2:
        _, v, x = cur
        dc.uf.addSum(v, x)
    if cur[0] == 3:
        _, v = cur
        res.append(dc.uf.getComponentSum(v))


dc.build()
dc.run(func)

print("\n".join(map(str, res)))

File: 22_专题/离线查询/并查集/Dynamic Graph Vertex Add Component Sum-连通分量和.go
// 离线动态连通性查询
// https://suisen-cp.github.io/cp-library-cpp/library/algorithm/offline_dynamic_connectivity_component_sum.hpp

package main

import (
	"bufio"
	"fmt"
	"math/bits"
	"os"
)

func main() {
	// 0 u v 连接u v (保证u v不连接)
	// 1 u v 断开u v  (保证u v连接)
	// 2 u x 将u的值加上x
	// 3 u 输出u所在连通块的值

	in := bufio.NewReader(os.Stdin)
	out := bufio.NewWriter(os.Stdout)
	defer out.Flush()

	var n, q int
	fmt.Fscan(in, &n, &q)
	sums := make([]int, n)
	for i := 0; i < n; i++ {
		fmt.Fscan(in, &sums[i])
	}
	queries := make([][]int, 0, q)
	for i := 0; i < q; i++ {
		var op, u, v, add int
		fmt.Fscan(in, &op)
		if op == 0 || op == 1 {
			fmt.Fscan(in, &u, &v)
			queries = append(queries, []int{op, u, v})
		} else if op == 2 {
			fmt.Fscan(in, &u, &add)
			queries = append(queries, []int{op, u, add})
		} else {
			fmt.Fscan(in, &u)
			queries = append(queries, []int{op, u})
		}
	}

	odc := NewOfflineDynamicConnectivity(n, q)
	for i, v := range sums {
		odc.Uf.Add(i, v) // 开始时每个点的权值
	}
	for i, query := range queries { // 处理边的添加和删除
		op := query[0]
		if op == 0 {
			u, v := query[1], query[2]
			odc.AddEdge(u, v, i)
		} else if op == 1 {
			u, v := query[1], query[2]
			odc.RemoveEdge(u, v, i)
		}
	}
	odc.Build()

	// 处理权值查询
	res := []int{}
	f := func(k int) {
		op := queries[k][0]
		if op == 2 {
			index, add := queries[k][1], queries[k][2]
			odc.Uf.Add(index, add)
		} else if op == 3 {
			index := queries[k][1]
			res = append(res, odc.Uf.Get(index))
		}
	}
	odc.Run(f)

	for _, v := range res {
		fmt.Fprintln(out, v)
	}
}

type S = int

type OfflineDynamicConnectivity struct {
	Uf         *UndoDSU
	Part       int // 连通分量数
	n, q, size int
	seg        [][]int
	edges      [][]int // (edgeId,time,edge)
	edgeId     map[int]int
	remain     map[int]struct{}
}

// 离线动态连通性查询
//  n: 顶点 0~n-1
//  q: 查询 0~q-1
func NewOfflineDynamicConnectivity(n, q int) *OfflineDynamicConnectivity {
	log := bits.Len(uint(q - 1))
	size := 1 << log
	seg := make([][]int, size<<1)
	uf := NewUndoDSU(n)
	return &OfflineDynamicConnectivity{
		Uf:     uf,
		Part:   n,
		n:      n,
		q:      q,
		size:   size,
		seg:    seg,
		edges:  [][]int{},
		edgeId: map[int]int{},
		remain: map[int]struct{}{},
	}
}

// 时刻time添加一条无向边(u,v)
func (o *OfflineDynamicConnectivity) AddEdge(u, v, time int) {
	if u > v {
		u, v = v, u
	}
	tuple := u*o.n + v
	o.remain[tuple] = struct{}{}
	o.edgeId[tuple] = time
}

// 时刻time删除一条无向边(u,v)
func (o *OfflineDynamicConnectivity) RemoveEdge(u, v, time int) {
	if u > v {
		u, v = v, u
	}
	tuple := u*o.n + v
	delete(o.remain, tuple)
	o.edges = append(o.edges, []int{o.edgeId[tuple], time, tuple})
}

func (o *OfflineDynamicConnectivity) Build() {
	for e := range o.remain {
		o.edges = append(o.edges, []int{o.edgeId[e], o.q, e})
	}
	for i := range o.edges {
		l, r, e := o.edges[i][0], o.edges[i][1], o.edges[i][2]
		o.add(l, r, e)
	}
}

// 执行所有查询
//  cb: func(k int): 当前处于第k个查询(0-based)
func (o *OfflineDynamicConnectivity) Run(cb func(k int)) {
	var dfs func(cur int)
	dfs = func(cur int) {
		if o.size+o.q <= cur {
			return
		}

		add := 0
		for _, e := range o.seg[cur] {
			ok := o.Uf.Union(e/o.n, e%o.n) // AddEdge
			if ok {
				add++
			}
		}
		o.Part -= add

		if cur >= o.size {
			cb(cur - o.size)
		} else {
			dfs(cur << 1)
			dfs(cur<<1 | 1)
		}

		for i := 0; i < len(o.seg[cur]); i++ {
			o.Uf.Undo() // RemoveEdge
		}
		o.Part += add
	}
	dfs(1)
}

func (o *OfflineDynamicConnectivity) add(l, r, e int) {
	l += o.size
	r += o.size
	for l < r {
		if l&1 == 1 {
			o.seg[l] = append(o.seg[l], e)
			l++
		}
		if r&1 == 1 {
			r--
			o.seg[r] = append(o.seg[r], e)
		}
		l >>= 1
		r >>= 1
	}
}

func NewUndoDSU(n int) *UndoDSU { return NewUndoDSUWithWeights(make([]S, n)) }
func NewUndoDSUWithWeights(weights []S) *UndoDSU {
	n := len(weights)
	ps, ws := make([]int, n), make([]S, n)
	for i := 0; i < n; i++ {
		ps[i] = -1
		ws[i] = weights[i] // e()
	}
	history := [][]int{}
	return &UndoDSU{parentSize: ps, weights: ws, history: history}
}

type UndoDSU struct {
	parentSize []int
	weights    []S
	history    [][]int
}

func (uf *UndoDSU) Add(index int, delta S) {
	x := index
	for x >= 0 {
		uf.weights[x] += delta // op()
		x = uf.parentSize[x]
	}
}

func (uf *UndoDSU) Get(index int) S { return uf.weights[uf.Find(index)] }

func (uf *UndoDSU) Undo() bool {
	if len(uf.history) == 0 {
		return false
	}
	lastY := uf.history[len(uf.history)-1]
	uf.history = uf.history[:len(uf.history)-1]
	lastX := uf.history[len(uf.history)-1]
	uf.history = uf.history[:len(uf.history)-1]
	y, py := lastY[0], lastY[1]
	x, px := lastX[0], lastX[1]
	if uf.parentSize[x] != px {
		uf.weights[x] -= uf.weights[y] // inv()
	}
	uf.parentSize[x] = px
	uf.parentSize[y] = py
	return true
}

func (uf *UndoDSU) Reset() {
	for len(uf.history) > 0 {
		uf.Undo()
	}
}

func (uf *UndoDSU) Find(x int) int {
	cur := x
	for uf.parentSize[cur] >= 0 {
		cur = uf.parentSize[cur]
	}
	return cur
}

func (uf *UndoDSU) Union(x, y int) bool {
	x, y = uf.Find(x), uf.Find(y)
	if -uf.parentSize[x] < -uf.parentSize[y] {
		x, y = y, x
	}
	uf.history = append(uf.history, []int{x, uf.parentSize[x]})
	uf.history = append(uf.history, []int{y, uf.parentSize[y]})
	if x == y {
		return false
	}
	uf.parentSize[x] += uf.parentSize[y]
	uf.parentSize[y] = x
	uf.weights[x] += uf.weights[y]
	return true
}

func (uf *UndoDSU) IsConnected(x, y int) bool { return uf.Find(x) == uf.Find(y) }
